### PR TITLE
feat(server): load face entities faster

### PR DESCRIPTION
**Pull request by l0nax**
The added indices are based on monitoring a server with 
- 6 users
- 120.000+ Pictures
- 500+ GiB Storage

Especially the query executed when loading the peoples' page gains a speedup of 4x.

### CODE REVIEW

File: server/src/infra/repositories/person.repository.ts [hunk 1]
```diff
@@ -84,6 +83,8 @@ export class PersonRepository implements IPersonRepository {
       .limit(500);
     if (!options?.withHidden) {
       queryBuilder.andWhere('person.isHidden = false');
+    } else {
+      queryBuilder.orderBy('person.isHidden', 'ASC');
```
Comment (martabal) [inline server/src/infra/repositories/person.repository.ts:87]: What do you want to do ?
With this change, people with names are not prioritized to be at the top of the list.

File: server/src/infra/migrations/1700747853368-AddAssetFaceIndicies.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { MigrationInterface, QueryRunner } from "typeorm"
+
+export class AddAssetFaceIndicies1700747853368 implements MigrationInterface {
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`CREATE INDEX IF NOT EXISTS IDX_asset_faces_on_assetId ON "asset_faces"("assetId");`);
+    await queryRunner.query(`CREATE INDEX IF NOT EXISTS IDX_asset_faces_on_personId_assetId ON "asset_faces" ("personId", "assetId");`);
```
Comment (zackpollard) [inline server/src/infra/migrations/1700747853368-AddAssetFaceIndicies.ts:7]: The above two indexes could be combined if the query WHERE clauses are ordered in the correct way. If we always query with personId after assetId then you could just have a single index. PostgreSQL will use the assetId,personId index for any query just involving assetId.

**Comment by zackpollard** [comment]
These indexes should be managed inside the typeorm schema. Take a look at the following docs from TypeORM, you can create the indexes using the schema and have it automatically generate the migration file for you. https://orkhan.gitbook.io/typeorm/docs/indices

File: server/src/infra/repositories/person.repository.ts (hunk 1 above)
Comment (l0nax) [reply server/src/infra/repositories/person.repository.ts:87]: If `withHidden` is set to true, all entries returned will have `isHidden` set to `FALSE` and thus sorting by `isHidden` does not change the returned items.

If I'm not mistaken, then the following two queries should be the same (with `withHidden = true`):
```sql
SELECT
  "person"."id" AS "person_id",
  "person"."createdAt" AS "person_createdAt",
  "person"."updatedAt" AS "person_updatedAt",
  "person"."ownerId" AS "person_ownerId",
  "person"."name" AS "person_name",
  "person"."birthDate" AS "person_birthDate",
  "person"."thumbnailPath" AS "person_thumbnailPath",
  "person"."faceAssetId" AS "person_faceAssetId",
  "person"."isHidden" AS "person_isHidden"
FROM "person" "person"
LEFT JOIN "asset_faces" "face"
  ON "face"."personId"="person"."id"
INNER JOIN "assets" "asset"
  ON "asset"."id"="face"."assetId"
    AND ("asset"."deletedAt" IS NULL)
WHERE
  "person"."ownerId" = $1
  AND "person"."isHidden" = FALSE
GROUP BY "person"."id"
HAVING
  "person"."name" != ''
  OR COUNT("face"."assetId") >= $2
ORDER BY
  "person"."isHidden" ASC,
  NULLIF("person"."name", '') IS NULL ASC,
  COUNT("face"."assetId") DESC,
  NULLIF("person"."name", '') ASC NULLS LAST
LIMIT 500
```
and 
```sql
SELECT
  "person"."id" AS "person_id",
  "person"."createdAt" AS "person_createdAt",
  "person"."updatedAt" AS "person_updatedAt",
  "person"."ownerId" AS "person_ownerId",
  "person"."name" AS "person_name",
  "person"."birthDate" AS "person_birthDate",
  "person"."thumbnailPath" AS "person_thumbnailPath",
  "person"."faceAssetId" AS "person_faceAssetId",
  "person"."isHidden" AS "person_isHidden"
FROM "person" "person"
LEFT JOIN "asset_faces" "face"
  ON "face"."personId"="person"."id"
INNER JOIN "assets" "asset"
  ON "asset"."id"="face"."assetId"
    AND ("asset"."deletedAt" IS NULL)
WHERE
  "person"."ownerId" = $1
  AND "person"."isHidden" = FALSE
GROUP BY "person"."id"
HAVING
  "person"."name" != ''
  OR COUNT("face"."assetId") >= $2
ORDER BY
  NULLIF("person"."name", '') IS NULL ASC,
  COUNT("face"."assetId") DESC,
  NULLIF("person"."name", '') ASC NULLS LAST
LIMIT 500
```

Comment (martabal) [reply server/src/infra/repositories/person.repository.ts:87]: No, this change the sql query : 
original :
```sql
SELECT
    "person"."id" AS "person_id",
    "person"."createdAt" AS "person_createdAt",
    "person"."updatedAt" AS "person_updatedAt",
    "person"."ownerId" AS "person_ownerId",
    "person"."name" AS "person_name",
    "person"."birthDate" AS "person_birthDate",
    "person"."thumbnailPath" AS "person_thumbnailPath",
    "person"."faceAssetId" AS "person_faceAssetId",
    "person"."isHidden" AS "person_isHidden"
FROM
    "person" "person"
LEFT JOIN
    "asset_faces" "face" ON "face"."personId" = "person"."id"
INNER JOIN
    "assets" "asset" ON "asset"."id" = "face"."assetId" AND ("asset"."deletedAt" IS NULL)
WHERE
    "person"."ownerId" = $1
GROUP BY
    "person"."id"
HAVING
    "person"."name" != '' OR COUNT("face"."assetId") >= $2
ORDER BY
    "person"."isHidden" ASC,
    NULLIF("person"."name", '') IS NULL ASC,
    COUNT("face"."assetId") DESC,
    NULLIF("person"."name", '') ASC NULLS LAST
LIMIT
    500;
```

your changes :
```sql
SELECT
    "person"."id" AS "person_id",
    "person"."createdAt" AS "person_createdAt",
    "person"."updatedAt" AS "person_updatedAt",
    "person"."ownerId" AS "person_ownerId",
    "person"."name" AS "person_name",
    "person"."birthDate" AS "person_birthDate",
    "person"."thumbnailPath" AS "person_thumbnailPath",
    "person"."faceAssetId" AS "person_faceAssetId",
    "person"."isHidden" AS "person_isHidden"
FROM
    "person" "person"
LEFT JOIN
    "asset_faces" "face" ON "face"."personId" = "person"."id"
INNER JOIN
    "assets" "asset" ON "asset"."id" = "face"."assetId" AND ("asset"."deletedAt" IS NULL)
WHERE
    "person"."ownerId" = $1
GROUP BY
    "person"."id"
HAVING
    "person"."name" != '' OR COUNT("face"."assetId") >= $2
ORDER BY
    "person"."isHidden" ASC
LIMIT 500;
```

Comment (l0nax) [reply server/src/infra/repositories/person.repository.ts]: Ah, I now understand why: `addOrderBy` appends entries to an `ORDER BY`.

I'm sorry for the confusion – I've reverted this change.

Comment (martabal) [reply server/src/infra/repositories/person.repository.ts]: Correct me if I'm wrong but you always want the hidden people to come after the non-hidden people. With your changes and `withHidden` set to false, if someone has more than 500 people hidden, the query will return nothing.

**Review by jrasm91 — COMMENTED** [review]
Thanks for the PR! Is there a way to have postgres show the slowest x queries? It would be cool to see what those are on a large instance.

**Review by martabal — APPROVED** [review]

**Comment by l0nax** [comment]
@jrasm91 there are many ways to monitor your PostgreSQL instance.

You could use, e.g., the [pg_stat_statements](https://www.postgresql.org/docs/current/pgstatstatements.html) extension.
Postgres can even generate execution plans for slow queries automatically (see https://www.postgresql.org/docs/current/auto-explain.html).

**Comment by davidpan** [comment]
> Thanks for the PR! Is there a way to have postgres show the slowest x queries? It would be cool to see what those are on a large instance.

For development and testing, you can use pg_stat_statements. 
link: https://www.postgresql.org/docs/current/pgstatstatements.html

**Review by jrasm91 — APPROVED** [review]
Thank you!

**Pull request merged**